src/chain_table.rs

```diff
@@ -21,121 +21,167 @@ use std::{
-        const TABLE_WIDTH: usize,
-        const DEM_KEY_LENGTH: usize,
-        DEM: Dem<DEM_KEY_LENGTH>,
-    >(
+    pub fn decrypt<const DEM_KEY_LENGTH: usize, DEM: Dem<DEM_KEY_LENGTH>>(
         kwi_value: &DEM::Key,
         ciphertext: &[u8],
     ) -> Result<Self, Error> {
-        if TABLE_WIDTH * BLOCK_LENGTH + DEM::ENCRYPTION_OVERHEAD != ciphertext.len() {
+        let max_ciphertext_length = 1 + TABLE_WIDTH * (1 + BLOCK_LENGTH) + DEM::ENCRYPTION_OVERHEAD;
+        if max_ciphertext_length != ciphertext.len() {
             return Err(Error::CryptoError(format!(
                 "invalid ciphertext length: given {}, should be {}",
                 ciphertext.len(),
-                TABLE_WIDTH * BLOCK_LENGTH + DEM::ENCRYPTION_OVERHEAD
+                max_ciphertext_length
             )));
         }
         let bytes = DEM::decrypt(kwi_value, ciphertext, None)?;
-        let mut de = Deserializer::new(&bytes);
-        let mut res = Vec::with_capacity(TABLE_WIDTH);
-        for _ in 0..TABLE_WIDTH {
-            let block = de.read::<Block<BLOCK_LENGTH>>()?;
-            // Blocks starting by `0` are padding.
-            // There cannot be any meaningful block after a padding block.
-            if block[0] == 0 {
-                break;
-            }
-            res.push(block);
-        }
-        Ok(Self(res))
+        Self::try_from_bytes(&bytes)
     }
 }
 
-impl<const BLOCK_LENGTH: usize> Deref for ChainTableValue<BLOCK_LENGTH> {
-    type Target = Vec<Block<BLOCK_LENGTH>>;
+impl<const TABLE_WIDTH: usize, const BLOCK_LENGTH: usize> Serializable
+    for ChainTableValue<TABLE_WIDTH, BLOCK_LENGTH>
+{
+    type Error = Error;
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+    fn length(&self) -> usize {
+        // The leading byte corresponds to the flag byte.
+        1 + TABLE_WIDTH * (1 + BLOCK_LENGTH)
     }
-}
 
-impl<const BLOCK_LENGTH: usize> DerefMut for ChainTableValue<BLOCK_LENGTH> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+    fn write(&self, ser: &mut Serializer) -> Result<usize, Self::Error> {
+        let mut flag = 0u8;
+        for i in 0..self.length {
+            // Set the corresponding bit to 1 in the flag byte.
+            if self.blocks[i].is_addition() {
+                flag += 1 << i;
+            }
+        }
+        let mut n = ser.write_array(&[flag])?;
+        for block in self.blocks {
+            n += ser.write_array(&[<u8>::from(block.prefix)])?;
+            n += ser.write_array(&block.data)?;
+        }
+        Ok(n)
     }
-}
-
-impl<const BLOCK_LENGTH: usize> IntoIterator for ChainTableValue<BLOCK_LENGTH> {
-    type IntoIter = std::vec::IntoIter<Block<BLOCK_LENGTH>>;
-    type Item = Block<BLOCK_LENGTH>;
 
-    fn into_iter(self) -> Self::IntoIter {
-        self.0.into_iter()
+    fn read(de: &mut Deserializer) -> Result<Self, Self::Error> {
+        let mut res = Self::default();
+        let mut flag = de.read_array::<1>()?[0];
```

Comment:
Flag could be not mutable here by using `flag & (1 << i)` to check if the bit is set (or something similar, doing it in my head ^^)

Reply:
Did something like that in f2c5ec6